onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -226,12 +227,258 @@ def compute_range(self):
 
         return self.calibrate_tensors_range
 
+class EntropyCalibrater(CalibraterBase):
+    def __init__(self, model, op_types_to_calibrate=[], augmented_model_path='augmented_model.onnx'):
+        '''
+        :param model: ONNX model to calibrate. It can be a ModelProto or a model path
+        :param op_types_to_calibrate: operator types to calibrate. By default, calibrate all the float32/float16 tensors.
+        :param augmented_model_path: save augmented model to this path.
+        '''
+        super(EntropyCalibrater, self).__init__(model, op_types_to_calibrate, augmented_model_path)
+        self.intermediate_outputs = []
+        self.calibrate_tensors_range = None
+        self.num_model_outputs = len(self.model.graph.output)
+        self.model_original_outputs = set(output.name for output in self.model.graph.output)
+        self.collector = None
+
+    def augment_graph(self):
+        '''
+        Adds ReduceMin and ReduceMax nodes to all quantization_candidates op type nodes in
+        model and ensures their outputs are stored as part of the graph output
+        :return: augmented ONNX model
+        '''
+        model = onnx_proto.ModelProto()
+        model.CopyFrom(self.model)
+        model = onnx.shape_inference.infer_shapes(model)
+        value_infos = {vi.name: vi for vi in model.graph.value_info}
+        value_infos.update({ot.name: ot for ot in model.graph.output})
+        value_infos.update({it.name: it for it in model.graph.input})
+        initializer = set(init.name for init in model.graph.initializer)
+
+        added_nodes = []
+        added_outputs = []
+        tensors_to_calibrate = set()
+        tensor_type_to_calibrate = set([TensorProto.FLOAT, TensorProto.FLOAT16])
+
+        for node in model.graph.node:
+            if len(self.op_types_to_calibrate) == 0 or node.op_type in self.op_types_to_calibrate:
+                for tensor_name in itertools.chain(node.input, node.output):
+                    if tensor_name in value_infos.keys():
+                        vi = value_infos[tensor_name]
+                        if vi.type.HasField('tensor_type') and (
+                                vi.type.tensor_type.elem_type in tensor_type_to_calibrate) and (
+                                    tensor_name not in initializer):
+                            tensors_to_calibrate.add(tensor_name)
```

Comment:
this is same with the MaxMin one. Could you extract the code and make it reusable

Reply:
Made it reusable.